core/trino-spi/src/main/java/io/trino/spi/predicate/SortedRangeSet.java

```diff
@@ -1042,6 +1046,87 @@ else if (thatNextRangeIndex == thatRangeCount) {
         return new SortedRangeSet(type, inclusive, blockBuilder.build(), unionIsDiscreteSet(that, resultRangeIndex > 0));
     }
 
+    // visible for testing
+    SortedRangeSet linearDiscreteSetUnion(SortedRangeSet that)
+    {
+        return linearDiscreteSetUnion(this, that);
+    }
+
+    private SortedRangeSet linearDiscreteSetUnion(SortedRangeSet leftRangeSet, SortedRangeSet rightRangeSet)
+    {
+        int leftRangeCount = leftRangeSet.getRangeCount();
+        int rightRangeCount = rightRangeSet.getRangeCount();
+
+        boolean[] inclusive = new boolean[2 * (leftRangeCount + rightRangeCount)];
+        BlockBuilder blockBuilder = type.createBlockBuilder(null, 2 * (leftRangeCount + rightRangeCount));
+
+        int resultRangeIndex = 0;
+        int leftNextRangeIndex = 0;
+        int rightNextRangeIndex = 0;
+
+        int currentUnionStart = 0;
+
+        // The value at leftNextRangeIndex is guaranteed to be smaller than or equals to the value
+        // at the corresponding index on the right side. This is maintained by comparing and swapping.
+        // Due to this property, we can safely add all values from the left side up to
+        // leftNextRangeIndex in bulk, thus preserving sorted order.
+        while (leftNextRangeIndex < leftRangeCount && rightNextRangeIndex < rightRangeCount) {
+            int compare = compareValues(
+                    comparisonOperator,
+                    leftRangeSet.sortedRanges,
+                    2 * leftNextRangeIndex,
+                    rightRangeSet.sortedRanges,
+                    2 * rightNextRangeIndex);
+
+            if (compare == 0) {
+                leftNextRangeIndex++;
+                rightNextRangeIndex++;
+            }
+            else if (compare < 0) {
+                leftNextRangeIndex++;
+            }
+            else {
+                int size = leftNextRangeIndex - currentUnionStart;
+                if (size > 0) {
+                    copyBlock(leftRangeSet, currentUnionStart * 2, blockBuilder, inclusive, resultRangeIndex * 2, size);
+                    resultRangeIndex += size;
+                    currentUnionStart = rightNextRangeIndex;
```

Comment:
add a comment why `currentUnionStart = rightNextRangeIndex` under `if (size > 0) {` 
OR just make it unconditional
OR add a check state `checkState(leftNextRangeIndex == 0)` because I think it's only case when `size == 0`

Reply:
Is it ok to remove the `if (size > 0)` ? 
I was thought the size should always greater than 0.

Reply:
Also I want to add a check if in `copyBlock` method, when the size is 0 return directly.

Reply:
Yup. You could make `copyBlock` work with size 0, but IMO it should be explicit check there